File: src/notify.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const config = require('./lib/config');
const telegram = require('./lib/telegram');
const formatter = require('./lib/formatter');
const sessions = require('./lib/sessions');
const tmux = require('./lib/tmux');
const transcript = require('./lib/transcript');
const { logEvent, logError } = require('./lib/logger');

const NOTIFY_LOG_PATH = path.join(__dirname, '..', '.notify-log.json');

function readStdin() {
  return new Promise((resolve) => {
    let data = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (chunk) => {
      data += chunk;
    });
    process.stdin.on('end', () => {
      resolve(data);
    });
    setTimeout(() => {
      resolve(data);
    }, 5000);
  });
}

async function main() {
  const input = await readStdin();

  let hookData;
  try {
    hookData = JSON.parse(input);
  } catch (_e) {
    logError(null, 'failed to parse stdin JSON');
    process.exit(0);
  }

  let cfg;
  try {
    cfg = config.loadConfig();
  } catch (_e) {
    logError(null, 'failed to load config');
    process.exit(0);
  }

  if (!cfg.chatId) {
    process.exit(0);
  }

  const { hook_event_name: eventName, notification_type: notificationType } =
    hookData;
  const sessionName = findSessionForCwd(hookData.cwd);

  if (eventName !== 'Notification') {
    process.exit(0);
  }
  if (notificationType === 'idle_prompt' && !cfg.notifyOn.idle) {
    logEvent(sessionName, 'idle notification filtered (disabled)');
    process.exit(0);
  }
  if (notificationType === 'permission_prompt' && !cfg.notifyOn.permission) {
    logEvent(sessionName, 'permission notification filtered (disabled)');
    process.exit(0);
  }

  if (notificationType === 'idle_prompt' && !cooldownExpired(cfg, 'lastIdle')) {
    logEvent(sessionName, 'idle notification filtered (cooldown)');
    process.exit(0);
  }

  if (notificationType === 'idle_prompt' && isUserTyping(hookData.cwd)) {
    logEvent(sessionName, 'idle notification filtered (user typing)');
    process.exit(0);
  }

  if (
    notificationType === 'permission_prompt' &&
    !cooldownExpired(cfg, 'lastPermission')
  ) {
    logEvent(sessionName, 'permission notification filtered (cooldown)');
    process.exit(0);
  }

  if (sessionName) {
    sessions.setActive(sessionName);
  }

  let screenContent = '';
  let isTranscriptContent = false;

  if (notificationType === 'permission_prompt' || notificationType === 'elicitation_dialog') {
    if (sessionName) {
      try {
        screenContent = tmux.capturePane(sessionName, 50).trim();
      } catch (_e) {
        // ignore capture failures
      }
    }
  } else {
    if (hookData.transcript_path) {
      try {
        const lastMsg = transcript.getLastAssistantMessage(
          hookData.transcript_path
        );
        if (lastMsg) {
          screenContent = lastMsg;
          isTranscriptContent = true;
        }
      } catch (_e) {
        // ignore
      }
    }

    if (!screenContent && sessionName) {
      try {
        screenContent = tmux.capturePane(sessionName, 50).trim();
      } catch (_e) {
        // ignore capture failures
      }
    }
  }

  const showHint = shouldShowHint();

  const message = formatter.formatNotification(
    hookData,
    sessionName,
    screenContent,
    showHint,
    isTranscriptContent
  );
  try {
    await telegram.sendMessageWithRetry(cfg.botToken, cfg.chatId, message);
    logEvent(sessionName, `${notificationType} notification sent`);
  } catch (err) {
    logError(sessionName, `telegram send failed: ${err.message}`);
    process.exit(0);
  }

  const log = loadLog();
  if (notificationType === 'idle_prompt') {
    log.lastIdle = Date.now();
  } else if (notificationType === 'permission_prompt') {
    log.lastPermission = Date.now();
  }
  saveLog(log);
}

function cooldownExpired(cfg, key) {
  const cooldownMs = (cfg.idleCooldown || 180) * 1000;
  const now = Date.now();
  const log = loadLog();
  const last = log[key] || 0;
  return now - last >= cooldownMs;
}

function loadLog() {
  try {
    return JSON.parse(fs.readFileSync(NOTIFY_LOG_PATH, 'utf8'));
  } catch (_e) {
    return { timestamps: [], lastIdle: 0, lastPermission: 0 };
  }
}

function saveLog(log) {
  try {
    fs.writeFileSync(NOTIFY_LOG_PATH, JSON.stringify(log));
  } catch (_e) {
    // ignore
  }
}

function shouldShowHint() {
  const now = Date.now();
  const window = 60000;
  const threshold = 3;

  const log = loadLog();
  if (!Array.isArray(log.timestamps)) log.timestamps = [];

  log.timestamps.push(now);
  log.timestamps = log.timestamps.filter((ts) => now - ts < window);
  saveLog(log);

  return log.timestamps.length >= threshold;
}

function isUserTyping(cwd) {
  try {
    const sessionName = findSessionForCwd(cwd);
    if (!sessionName) return false;
    const lastLine = tmux.capturePane(sessionName, 1).trim();
    // Check if there's text after the ❯ prompt (user is drafting input)
    const match = lastLine.match(/❯\s?(.+)/);
    return match !== null && match[1].trim().length > 0;
  } catch (_e) {
    return false;
  }
}

function findSessionForCwd(cwd) {
  if (!cwd) return null;
  const state = sessions.list();
  const names = Object.keys(state.sessions);
  for (const name of names) {
    if (state.sessions[name].cwd === cwd) {
      return name;
    }
  }
  return state.active;
}

main().catch((err) => {
  logError(null, `uncaught error: ${err.message}`);
  process.exit(0);
});

File: internal/notify/notify.go
package notify

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"time"

	"claude-tg/internal/config"
	"claude-tg/internal/formatter"
	"claude-tg/internal/sessions"
	"claude-tg/internal/telegram"
	"claude-tg/internal/tmux"
	"claude-tg/internal/transcript"
)

var typingRegex = regexp.MustCompile(`❯\s?(.+)`)

type HookData struct {
	EventName        string `json:"hook_event_name"`
	NotificationType string `json:"notification_type"`
	Message          string `json:"message"`
	Title            string `json:"title"`
	Cwd              string `json:"cwd"`
	TranscriptPath   string `json:"transcript_path"`
}

type NotifyLog struct {
	LastIdle      int64   `json:"lastIdle"`
	LastPermission int64  `json:"lastPermission"`
	Timestamps    []int64 `json:"timestamps"`
}

func logPath() string {
	return filepath.Join(config.ProjectRoot(), ".notify-log.json")
}

func loadLog() *NotifyLog {
	data, err := os.ReadFile(logPath())
	if err != nil {
		return &NotifyLog{}
	}
	var log NotifyLog
	if err := json.Unmarshal(data, &log); err != nil {
		return &NotifyLog{}
	}
	return &log
}

func saveLog(log *NotifyLog) {
	data, _ := json.Marshal(log)
	os.WriteFile(logPath(), data, 0644)
}

func cooldownExpired(cfg *config.Config, key string) bool {
	log := loadLog()
	cooldownMs := int64(cfg.IdleCooldown) * 1000
	now := time.Now().UnixMilli()
	var last int64
	if key == "lastIdle" {
		last = log.LastIdle
	} else {
		last = log.LastPermission
	}
	return now-last >= cooldownMs
}

func shouldShowHint() bool {
	nlog := loadLog()
	now := time.Now().UnixMilli()
	window := int64(60000) // 60 seconds

	// Clean old timestamps
	var recent []int64
	for _, ts := range nlog.Timestamps {
		if now-ts < window {
			recent = append(recent, ts)
		}
	}
	recent = append(recent, now)
	nlog.Timestamps = recent
	saveLog(nlog)

	return len(recent) >= 3
}

func isUserTyping(cwd string) bool {
	sessionName := sessions.FindByCwd(cwd)
	if sessionName == "" {
		return false
	}
	content, err := tmux.CapturePane(sessionName, 1, "")
	if err != nil {
		return false
	}
	matches := typingRegex.FindStringSubmatch(content)
	if matches != nil && len(matches) > 1 {
		return len(matches[1]) > 0
	}
	return false
}

func isBotRunning() bool {
	pidPath := filepath.Join(config.ProjectRoot(), ".bot.pid")
	data, err := os.ReadFile(pidPath)
	if err != nil {
		return false
	}
	var pid int
	if _, err := fmt.Sscanf(strings.TrimSpace(string(data)), "%d", &pid); err != nil {
		return false
	}
	return syscall.Kill(pid, 0) == nil
}

func Run() {
	// Always exit 0 to never block Claude
	defer func() {
		if r := recover(); r != nil {
			debugLog("panic: %v", r)
			os.Exit(0)
		}
	}()

	if err := run(); err != nil {
		debugLog("notify error: %v", err)
		os.Exit(0)
	}
	os.Exit(0)
}

func debugLog(format string, args ...interface{}) {
	f, err := os.OpenFile(filepath.Join(config.ProjectRoot(), "notify-debug.log"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	fmt.Fprintf(f, "%s %s\n", time.Now().Format(time.RFC3339), fmt.Sprintf(format, args...))
}

func run() error {
	// Read stdin with timeout
	stdinCh := make(chan []byte, 1)
	go func() {
		data, err := io.ReadAll(os.Stdin)
		if err != nil {
			stdinCh <- nil
			return
		}
		stdinCh <- data
	}()

	var stdinData []byte
	select {
	case stdinData = <-stdinCh:
	case <-time.After(5 * time.Second):
		return fmt.Errorf("stdin timeout")
	}

	if len(stdinData) == 0 {
		return fmt.Errorf("empty stdin")
	}

	var hookData HookData
	if err := json.Unmarshal(stdinData, &hookData); err != nil {
		return fmt.Errorf("invalid JSON: %w", err)
	}

	cfg, err := config.LoadConfig()
	if err != nil {
		return err
	}

	if cfg.ChatID == "" {
		return fmt.Errorf("no chat ID configured")
	}

	if !isBotRunning() {
		return fmt.Errorf("bot not running")
	}

	// Only process Notification events
	if hookData.EventName != "Notification" {
		return nil
	}

	// Check notification type settings
	switch hookData.NotificationType {
	case "idle_prompt":
		if !cfg.NotifyOnIdle {
			return nil
		}
		if !cooldownExpired(cfg, "lastIdle") {
			return nil
		}
		if isUserTyping(hookData.Cwd) {
			return nil
		}
	case "permission_prompt", "elicitation_dialog":
		if !cfg.NotifyOnPermission {
			return nil
		}
		// No cooldown for permission prompts — each one requires user action
	default:
		return nil
	}

	// Find session and set active
	sessionName := sessions.FindByCwd(hookData.Cwd)
	if sessionName != "" {
		sessions.SetActive(sessionName)
	}

	// Capture content
	var screenContent string
	isTranscript := false
	if hookData.NotificationType == "permission_prompt" || hookData.NotificationType == "elicitation_dialog" {
		if sessionName != "" {
			screenContent, _ = tmux.CapturePane(sessionName, 200, "")
		}
	} else {
		// Idle: prefer transcript
		if hookData.TranscriptPath != "" {
			content := transcript.GetLastAssistantMessage(hookData.TranscriptPath)
			if content != "" {
				screenContent = content
				isTranscript = true
			}
		}
		if screenContent == "" && sessionName != "" {
			screenContent, _ = tmux.CapturePane(sessionName, 50, "")
		}
	}

	showHint := shouldShowHint()

	message := formatter.FormatNotification(
		hookData.NotificationType,
		sessionName,
		hookData.Title,
		hookData.Message,
		screenContent,
		showHint,
		isTranscript,
	)

	client := telegram.NewClient(cfg.BotToken)
	client.SendMessageWithRetry(cfg.ChatID, message, "HTML")

	// Update throttle log
	nlog := loadLog()
	now := time.Now().UnixMilli()
	if hookData.NotificationType == "idle_prompt" {
		nlog.LastIdle = now
	} else {
		nlog.LastPermission = now
	}
	saveLog(nlog)

	return nil
}
